New input for Crowdstrike Falcon events

tonymeehan opened the pull request:
## What does this PR do?

This adds a new input for Crowdstrike Falcon events forwarded by Crowdstrike's [SIEM forwarder](https://www.crowdstrike.com/blog/tech-center/integrate-with-your-siem/). This input uses the default JSON output format from the SIEM forwarder.

## Why is it important?

We've had several users asks us for adding support for Crowdstrike Falcon in ECS. We strive to be data agnostic to help enable our SOC users aggregate all of their data in our SIEM. Using some log data provided by our users, I've put this new module together to ingest the data and convert many fields into ECS (including categorization).

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works

## How to test this PR locally

```
mage update build integTest
```

## Screenshots

![crowdstrike mov](https://user-images.githubusercontent.com/19255656/76566501-4cc9ac00-6483-11ea-9e37-44a1ec8a1eff.gif)

## Proof this works

```
[success] 0.85% test_xpack_modules.XPackTest.test_fileset_file_008_crowdstrike: 2.7708s
----------------------------------------------------------------------
Ran 135 tests in 326.929s

OK
```

I also ran ```make docs``` to validate the documentation looked good.

## Other comments

I'll squash before merging.

elasticmachine commented:
Pinging @elastic/siem (Team:SIEM)

tonymeehan commented:
Looking into CI errors. Need to run ```make check``` and ```mage fmt update```

leehinman commented on x-pack/filebeat/module/crowdstrike/falcon/test/falcon-events.log-expected.json:
```diff
@@ -0,0 +1,572 @@
+        "event.action": "remote_response_session_start_event",
+        "event.type": "info",
+        "host.name": "hostnameofmachine",
+        "user.name": "first.last@company.com",
```
this looks like user.email, will these always be emails?  maybe switch on the presence of@ for email vs name?

tonymeehan replied:
This is an interesting question that I don't know the answer to since we don't have access to the Falcon platform. This is the username of the Crowdstrike Falcon user that is initiating a remote session to the endpoint. I'm not sure if it's guaranteed to be an email. But it's definitely meant to be the username since the original field is UserName (crowdstrike.event.UserName).

webmat replied:
Yeah it's common to see emails in "user id" or "user name" fields in different systems.

Similar to the point I made about `UserId` vs `UserName` in a comment on `AuthActivityAuditEvent`, I'd suggest filling our `user.name` with this email (as you did here), to try to follow their semantics as closely as possible.

Then perhaps do a check and also fill the `user.email` field with the same value if the value in question looks like an email. I wouldn't overthink the check, the presence of `@` in the field is probably enough.

leehinman commented on x-pack/filebeat/module/crowdstrike/falcon/test/falcon-events.log-expected.json:
```diff
@@ -0,0 +1,572 @@
+        "crowdstrike.event.UserIp": "10.10.0.8",
+        "crowdstrike.event.OperationName": "streamStarted",
+        "crowdstrike.event.ServiceName": "Crowdstrike Streaming API",
+        "crowdstrike.event.Success": true,
```
should we map this to event.outcome?

tonymeehan replied:
Good idea.

leehinman commented on x-pack/filebeat/module/crowdstrike/falcon/test/falcon-events.log-expected.json:
```diff
@@ -0,0 +1,572 @@
+        "crowdstrike.event.UTCTimestamp": 1581546248,
+        "crowdstrike.event.UserId": "chris@company.com",
+        "crowdstrike.event.UserIp": "192.168.6.13",
+        "crowdstrike.event.OperationName": "update_group",
```
this looks like event.category should contain "iam"

leehinman commented on x-pack/filebeat/module/crowdstrike/falcon/test/falcon-events.log:
```diff
@@ -0,0 +1,345 @@
+        "FileName": "explorer.exe",
+        "FilePath": "\\Device\\HarddiskVolume1\\Windows",
+        "CommandLine": "C:\\Windows\\Explorer.EXE",
+        "SHA256String": "6a671b92a69755de6fd063fcbe4ba926d83b49f78c42dbaeed8cdb6bbc57576a",
```
should these be added to file.hash.sha256 & file.hash.md5 ?

tonymeehan commented on x-pack/filebeat/module/crowdstrike/falcon/config/pipeline.js:
```diff
@@ -0,0 +1,132 @@
+
+        evt.Put("agent.type", "falcon")
+        evt.Put("event.kind", "event")
+        evt.Put("host.name", "")
```
This is a good question. I was unable to delete the field. It's added later by filebeat. So it's misleading in the UI to see an event that came from some other endpoint associated with the filebeat agent. I thought it would be better to show an empty field in the UI to not mislead a user. But I'm 100% open to suggestions.

webmat replied:
Oh I thought I had amended that comment. I realized after the fact that in your description you say this log would be collected centrally, so disregard my comment on "running Filebeat on the host".

But we'll have to figure out what to do, to fill `host.name` though.

Filling it with the server's name for their server audit logs is the right thing to do as this is where these events are being reported. Then overwriting it with the event's computername for endpoint events is the right thing to do as well.

tonymeehan replied:
> Broader discussion point: as I note in one of my comments, this seems to be a log that mixes their service's audit logs with actual endpoint detection events. Both are worth capturing, of course. But I wonder if we should find a way to separate them out in two different `event.dataset`, as their meaning & semantics are pretty different.

This is a good suggestion. Curious for suggestions, here.

Maybe ```crowdstrike.falcon.events``` and ```crowdstrike.falcon.auditlogs```?

webmat replied:
My understanding is that it's usually tied to each subdirectory of a given module (e.g. Apache HTTPD has two sister directories under https://github.com/elastic/beats/tree/master/filebeat/module/apache, one for "apache.access" logs, one for "apache.error" logs). So not sure if it's going to be enough to  simply overwrite `event.dataset` with two different values from a single pipeline, but it's worth a try.

Now for the values in `event.dataset`, ECS doesn't mandate anything. But the Filebeat convention is usually two levels. Perhaps something like "crowdstrike.endpoint" and "crowdstrike.server"? Not sure how important "Falcon" is in the branding, or if they have many other products or what. Perhaps "crowdstrike.falcon.endpoint" and "crowdstrike.falcon.server" would be fine as well. I'm not aware of anything that has a hard dependency on this convention.

Auditbeat for example fills `event.dataset` with a single word, and that hasn't been a problem, afaik.

leehinman replied:
> This is a good question. I was unable to delete the field. It's added later by filebeat. So it's misleading in the UI to see an event that came from some other endpoint associated with the filebeat agent. I thought it would be better to show an empty field in the UI to not mislead a user. But I'm 100% open to suggestions.

Just an FYI, if you had done the Delete in an Elasticsearch pipeline it would have worked, that happens in ES after all the Filebeat processing.

tonymeehan replied:
Yep. I tried to intentionally do nothing in ES to keep things simple for any users that might be using Logstash.

tonymeehan commented on x-pack/filebeat/module/crowdstrike/falcon/config/pipeline.js:
```diff
@@ -0,0 +1,132 @@
+ 
+                break;
+
+            case "RemoteResponseSessionStartEvent":
```
Oh good call.

webmat commented on x-pack/filebeat/module/crowdstrike/falcon/config/pipeline.js:
```diff
@@ -0,0 +1,132 @@
+                break;
+
+            case "AuthActivityAuditEvent":
+                evt.Put("user.name", evt.Get("crowdstrike.event.UserId"))
```
Since Crowdstrike seems to alternate between providing a only `UserId` or only a `UserName`, I'd suggest to be strict in populating

* `user.id` with `UserId`
* `user.name` with `UserName`

Then if we want to conditionally backfill `user.name` with `UserId` when `user.name` is not present, that may make sense.

I wonder if there's a relation with their API service logs always filling `UserId` and the endpoint events filling `UserName`?

webmat commented on x-pack/filebeat/module/crowdstrike/falcon/test/falcon-events.log-expected.json:
```diff
@@ -0,0 +1,572 @@
+        "crowdstrike.event.Objective": "Falcon Detection Method",
+        "crowdstrike.event.ProcessId": 38684386611,
+        "crowdstrike.event.DetectDescription": "Terminated a process related to the deletion of backups, which is often indicative of ransomware activity.",
+        "crowdstrike.event.PatternDispositionDescription": "Prevention, process killed.",
```
This field sounds perfect to populate `event.action`. Unfortunately it sounds like it's not reliably populated. It's only on 1 of the 15 events

tonymeehan replied:
I think it's okay to use it here for this event type. Made the change.

webmat commented on x-pack/filebeat/module/crowdstrike/falcon/test/falcon-events.log-expected.json:
```diff
@@ -0,0 +1,572 @@
+        "crowdstrike.event.DetectId": "ldt:ec86abd353824e96765ecbe18eb4f0b4:38655257584",
+        "crowdstrike.event.MACAddress": "xx-xx-xx-xx-xx",
+        "crowdstrike.event.ProcessStartTime": 1536846339,
+        "crowdstrike.event.DetectName": "Process Terminated",
```
Would it make sense to map this one to `rule.name`? This one doesn't sound like some intricate rule, and we don't have other samples filling this one. But based on the definition of the field:

> Name of the detection.

Same for `DetectDescription` => `rule.description`

tonymeehan replied:
I'm good with this. Change made.

webmat commented on x-pack/filebeat/module/crowdstrike/falcon/config/pipeline.js:
```diff
@@ -0,0 +1,132 @@
+            { from: "crowdstrike.event.ProcessId", to: "process.pid" },
+
+            // UserActivityAuditEvent and AuthActivityAuditEvent
+            { from: "crowdstrike.event.UserIp", to: "source.ip" },
```
If we aren't 100% of this always containing a valid IP address, we should fill `source.address` (a `keyword` field, so ingestion will never fail), and conditionally fill `source.ip` (an `ip` field), when the `.address` field is actually an IP.

You can check how other modules do it. Although some like Suricata do it using a combination of Beat processors, and not the Javascript processor. Are you aware of a good example of doing it in the script processor somewhere, @leehinman?

webmat replied:
Or is this `Convert` processor in the script processor actually calling exactly the same code? I guess so. So perhaps you can repro the same behaviour by passing both `type: ip` and the `ignore_failure` param to that Convert processor...

tonymeehan replied:
Yeah, I'll add ignore_failure: true and type: ip. Good suggestions.

tonymeehan commented on x-pack/filebeat/module/crowdstrike/falcon/config/pipeline.js:
```diff
@@ -0,0 +1,132 @@
+                var tactic = evt.Get("crowdstrike.event.Tactic").toLowerCase()
+                var technique = evt.Get("crowdstrike.event.Technique").toLowerCase()
+                  
+                evt.Put("event.action", tactic + "_" + technique)
```
Yep, that's right. I'm fine changing the action to a verb for events.

leehinman commented on x-pack/filebeat/module/crowdstrike/falcon/config/pipeline.js:
```diff
@@ -0,0 +1,132 @@
+                  
+                evt.Put("event.action", tactic + "_" + technique)
+                evt.Put("event.kind", "alert")
+                evt.Put("event.type", "info")
```
Append in the Filebeat script processor will make the field an array if there are 2 or more items, if there is just one item it isn't an array.  I got bit by this in surricata.

tonymeehan replied:
Oh great catch, thank you.

tsg commented on x-pack/filebeat/module/crowdstrike/falcon/config/falcon.yml:
```diff
@@ -0,0 +1,19 @@
+type: log
```
I wonder if it's worth also supporting the syslog input here? https://www.crowdstrike.com/blog/tech-center/integrate-with-your-siem/ seems to suggest that it's possible for the Connector to send date directly to a syslog server?

tonymeehan replied:
It would be easy to add the config option for syslog and maybe not make it the default choice. But my concern is we can't test it to validate that the output is identical to the file output. I'm on the fence but can be convinced to add it if it's not the default choice.

tsg replied:
I see, makes sense. I think we can leave it out and add it if it's being requested. Then we have a way to validate it.

andrewkroh commented:
It looks like CI is much more happy now. The failures are from flakey tests and appear unrelated to your changes.

tonymeehan commented:
> It looks like CI is much more happy now. The failures are from flakey tests and appear unrelated to your changes.

Yep, thanks @andrewkroh. I was just rebuilding to see if that was the case. I'll start working on the recommendations tonight.

andrewkroh commented on x-pack/filebeat/module/crowdstrike/_meta/fields.yml:
```diff
@@ -0,0 +1,42 @@
+                Offset number that tracks the location of the event in stream. This is used to identify unique detection events.
+
+            - name: customerIDString
+              type: text
```
In most cases where `text` has been used would probably be better as `keyword`. Then if there were a case where `text` was required we could add a `.text` multi-field. Checkout https://www.elastic.co/guide/en/ecs/current/ecs-conventions.html#_ecs_convention_for_indexing_text_fields.

andrewkroh commented on x-pack/filebeat/module/crowdstrike/falcon_audit/config/pipeline.js:
```diff
@@ -0,0 +1,145 @@
+        if (outcome == true) {
+            evt.Put("event.outcome", "success")
+        }
+        else if (outcome == false) {
```
```suggestion
        else if (outcome === false) {
```

I think a strict-equality check would be better in most cases. Consider the case where `outcome = ""`, you'd get `failure` rather than `unknown`.

andrewkroh commented on x-pack/filebeat/module/crowdstrike/falcon_endpoint/_meta/fields.yml:
```diff
@@ -0,0 +1,161 @@
+    - name: CommandLine
+      type: text
+      description: >
+        Executable path with comand line arguments.
```
```suggestion
        Executable path with command line arguments.
```

andrewkroh commented on x-pack/filebeat/module/crowdstrike/falcon_endpoint/config/pipeline.js:
```diff
@@ -0,0 +1,126 @@
+                evt.Put("event.severity", evt.Get("crowdstrike.event.Severity"))
+                evt.Put("message", evt.Get("crowdstrike.event.DetectDescription"))
+                evt.Put("process.name", evt.Get("crowdstrike.event.FileName"))
+                evt.Put("process.executable", evt.Get("crowdstrike.event.CommandLine"))
```
Do you know if this value can contain arguments? The sample event doesn't have any args, but the name suggests that it may. I would suggest using `process.command_line` for this value. Then derive `process.executable` from the command_line by taking the first argument (like split(" ")[0]).

tonymeehan commented on x-pack/filebeat/module/crowdstrike/_meta/config.yml:
```diff
@@ -0,0 +1,8 @@
+- module: crowdstrike
+  # All logs
+  falcon:
+    enabled: true
```
I did some testing. This looks like it needs to be true. You can see this in other modules, like [cisco](https://github.com/elastic/beats/blob/master/x-pack/filebeat/module/cisco/_meta/config.yml#L3) or [googlecloud](https://github.com/elastic/beats/blob/master/x-pack/filebeat/module/googlecloud/_meta/config.yml#L3). I did a quick test and set these to false, then ran ```./filebeat modules enable crowdstrike```. It didn't collect the data until I changed ```modules.d/crowdstrike.yml``` to set enabled to true. Leaving it as is so that when a user runs ```./filebeat modules enable crowdstrike``` this works without having to edit enabled to set it to true.

tonymeehan commented:
Thanks for the feedback.

**PR Feedback Addressed**

- [X] Split up the filesets into two components, falcon_endpoint and falcon_audit.
- [X] Added additional ECS 1.5 fields
- [X] Fixed some bugs (e.g. event.type and event.category are array fields)

**PR Feedback Not Addressed**

- [ ] Geo IP -- two reasons -- most endpoint IPs will be private, geo data won't make sense here. And the the other reason was I wanted to avoid installing an ingest pipeline for our users that have Logstash setup between their Beats and the stack.
- [ ] Setting enabled to false in config.yml (rationale explained above)

<img width="1663" alt="Screen Shot 2020-03-26 at 9 09 58 AM" src="https://user-images.githubusercontent.com/19255656/77669849-c9688a00-6f5b-11ea-9cee-6a4efe783bc2.png">

<img width="1676" alt="Screen Shot 2020-03-26 at 9 10 37 AM" src="https://user-images.githubusercontent.com/19255656/77669854-ca012080-6f5b-11ea-9e5e-acf9d4bea0e3.png">

andrewkroh commented on x-pack/filebeat/module/crowdstrike/falcon_audit/manifest.yml:
```diff
@@ -0,0 +1,8 @@
+var:
+  - name: paths
+    default:
+      - /var/log/crowdstrike/falconhoseclient/output
```
I just realized that both of these filesets read from the same file. Filebeat won't allow two concurrent readers of the same file because it persists the read offset of the file just once in its registry. 

So would you ever expect both filesets enabled the same machine? If the goal of splitting them is to minimize processing of certain events that we don't care about then perhaps a good solution is to have a single fileset with a config option to control what event types it consumes.

You could pass a parameter into your module and then into your script.

```
processors:
- script:
    lang: javascript
    params:
      mode: [audit] # or [audit, endpoint]
```

tonymeehan replied:
We should make it one fileset/pipeline then. It will always be the same output file. I was already a bit nervous about separating it like this given the additional overhead of doubling the JSON decoding/parsing. The purpose as suggested by @webmat was to semantically separate the two types of event data. Let me see if I can set the dataset in the Javascript pipeline and accomplish the same result there.

tonymeehan commented:
jenkins retest this please

tonymeehan commented:
I've been able to address most of the PR feedback (thanks again, everyone). I think we're ready for some LGTMs unless anyone has any additional feedback/changes.

andrewkroh commented on x-pack/filebeat/module/crowdstrike/_meta/docs.asciidoc:
```diff
@@ -0,0 +1,79 @@
+
+== Crowdstrike module
+
+This is the filebeat module for the Crowdstrike Falcon using the Falcon https://www.crowdstrike.com/blog/tech-center/integrate-with-your-siem[SIEM Connector]. This module collects this data, converts it to ECS, and ingests it to view in the SIEM. By default, the Falcon SIEM connector outputs JSON formated Falcon Streaming API event data in JSON format.
```
```suggestion
This is the filebeat module for the Crowdstrike Falcon using the Falcon https://www.crowdstrike.com/blog/tech-center/integrate-with-your-siem[SIEM Connector]. This module collects this data, converts it to ECS, and ingests it to view in the SIEM. By default, the Falcon SIEM connector outputs JSON formatted Falcon Streaming API event data.
```

andrewkroh commented on x-pack/filebeat/module/crowdstrike/_meta/docs.asciidoc:
```diff
@@ -0,0 +1,79 @@
+
+This is the filebeat module for the Crowdstrike Falcon using the Falcon https://www.crowdstrike.com/blog/tech-center/integrate-with-your-siem[SIEM Connector]. This module collects this data, converts it to ECS, and ingests it to view in the SIEM. By default, the Falcon SIEM connector outputs JSON formated Falcon Streaming API event data in JSON format.
+
+This module segments events forwarded by the Falcon SIEM connector into two datasets for endpoint data and Falcon platform audit data.
```
```suggestion
```

tonymeehan replied:
I was going to keep this since we're still doing this (two datasets), just from the same fileset. That okay?

andrewkroh replied:
Oh, that's interesting. I didn't notice that the module was setting the `event.dataset` values. We've never had a case AFAIK where a fileset produced an `event.dataset` value other than the default of `{module}.{fileset}`. I don't think this will cause any issues. I think it would be good to callout the event.dataset values that it produces in this paragraph.

andrewkroh commented on x-pack/filebeat/module/crowdstrike/_meta/docs.asciidoc:
```diff
@@ -0,0 +1,79 @@
+include::../include/config-option-intro.asciidoc[]
+
+[float]
+==== `falcon_endpoint` fileset settings
```
```suggestion
==== `falcon` fileset settings
```

leehinman commented:
Changes look really good.

Looks like you need to run "mage fmt update" and push changes though.

tonymeehan commented:
Thanks for all the help @leehinman @andrewkroh @webmat and @tsg!
